cmd/resin-provision/main.go

```diff
@@ -4,139 +4,155 @@ import (
-	i := bufio.NewReader(os.Stdin)
-	in, err := i.ReadString('\n')
-	if err != nil {
-		return
+	oneshotCmd := &cobra.Command{
+		Use:      "oneshot",
+		Short:    "One-shot mode",
+		PreRunE:  validate,
+		RunE:     oneshot,
+		PostRunE: success,
 	}
-	input = strings.Trim(in, "\n")
-	return
-}
 
-func contains(s []string, e string) bool {
-	for _, a := range s {
-		if strings.Compare(a, e) == 0 {
-			return true
-		}
+	interactiveCmd := &cobra.Command{
+		Use:      "interactive",
+		Short:    "Interactive mode",
+		PreRunE:  validate,
+		RunE:     interactive,
+		PostRunE: success,
 	}
-	return false
-}
 
-func prompt(options []string, promptMessage string) (input string, err error) {
-	for {
-		fmt.Printf(promptMessage)
-		if input, err = readInput(); err != nil {
-			return
-		} else if len(options) == 0 {
-			return
-		} else if contains(options, input) {
-			return
-		}
+	statusCmd := &cobra.Command{
+		Use:   "status",
+		Short: "Check whether this device is provisioned",
+		RunE:  status,
 	}
+
+	rootCmd.AddCommand(oneshotCmd)
+	rootCmd.AddCommand(interactiveCmd)
+	rootCmd.AddCommand(statusCmd)
+
+	rootCmd.PersistentFlags().StringVarP(&domain, "domain", "d", defaults.RESIN_DOMAIN, "Domain the device will provision to")
+	rootCmd.PersistentFlags().StringVarP(&configPath, "config", "c", defaults.CONFIG_PATH, "Supervisor config path")
+	rootCmd.PersistentFlags().BoolVarP(&dryRun, "dryrun", "r", false, "Dry run (do not provision")
+	oneshotCmd.Flags().StringVarP(&application, "application", "a", "", "Application the device will provision to")
+	oneshotCmd.Flags().StringVarP(&email, "email", "e", "", "Email addres of the account the device will provision to")
+	oneshotCmd.Flags().StringVarP(&password, "password", "p", "", "Password of the account the device will provision to")
+	oneshotCmd.Flags().StringVarP(&token, "token", "t", "", "User AUTH token")
+
+	rootCmd.Execute()
 }
 
-func login() (token string, err error) {
-	fmt.Println("Logging in...")
-	for {
-		if email, e := prompt(nil, "email: "); err != nil {
-			return "", e
-		} else {
-			fmt.Printf("password: ")
-			if p, e := gopass.GetPasswdMasked(); e != nil {
-				return "", e
-			} else {
-				password := string(p)
-				if token, e := resin.Login("https://api."+domain, email, password); e != nil {
-					return "", e
-				} else if token != "" {
-					return token, nil
-				} else {
-					fmt.Println("Wrong email or password, please try again.")
-				}
+func validate(cmd *cobra.Command, args []string) error {
+	if cmd.Use == "oneshot" {
+		if application == "" {
+			return errors.New("Application is required")
+		} else if token == "" {
+			if email == "" {
+				return errors.New("Email is required")
+			} else if password == "" {
+				return errors.New("Password is required")
 			}
 		}
 	}
+
+	if status, err := provisioner.Status(configPath); err != nil {
+		return err
+	} else if status {
+		return errors.New("Your device is already provisioned")
```

Comment:
The binary was exiting with a zero because I didn't handle the result of `rootCmd.Execute` - this is fixed now.

I have moved the status check out of the validate function and into the provisioning commands (`oneshot` and `interactive`) which enables the binary to return 0 if the device is already provisioned.